Readme.md
# retext-pos [![Build Status](https://img.shields.io/travis/wooorm/retext-pos.svg?style=flat)](https://travis-ci.org/wooorm/retext-pos) [![Coverage Status](https://img.shields.io/coveralls/wooorm/retext-pos.svg?style=flat)](https://coveralls.io/r/wooorm/retext-pos?branch=master)

**[Retext](https://github.com/wooorm/retext "Retext")** POS (part-of-speech) tagger using [dariusk/pos-js](https://github.com/dariusk/pos-js).

## Installation

[npm](https://docs.npmjs.com/cli/install):

```bash
$ npm install retext-pos
```

[Component.js](https://github.com/componentjs/component):

```bash
$ component install wooorm/retext-pos
```

[Bower](http://bower.io/#install-packages):

```bash
$ bower install retext-pos
```

[Duo](http://duojs.org/#getting-started):

```javascript
var pos = require('wooorm/retext-pos');
```

## Usage

```javascript
var Retext = require('retext');
var visit = require('retext-visit');
var inspect = require('retext-inspect');
var pos = require('retext-pos');

var retext = new Retext()
    .use(visit)
    .use(inspect)
    .use(pos);

retext.parse(
    'I went to the store, to buy 5.2 gallons of milk.',
    function (err, tree) {
        tree.visit(tree.WORD_NODE, function (node) {
            console.log(node);
        });
        /*
         * WordNode[1] [data={"partOfSpeech":"NN"}]
         * └─ TextNode: 'I'
         *
         * WordNode[1] [data={"partOfSpeech":"VBD"}]
         * └─ TextNode: 'went'
         *
         * WordNode[1] [data={"partOfSpeech":"TO"}]
         * └─ TextNode: 'to'
         *
         * WordNode[1] [data={"partOfSpeech":"DT"}]
         * └─ TextNode: 'the'
         *
         * WordNode[1] [data={"partOfSpeech":"NN"}]
         * └─ TextNode: 'store'
         *
         * WordNode[1] [data={"partOfSpeech":"TO"}]
         * └─ TextNode: 'to'
         *
         * WordNode[1] [data={"partOfSpeech":"VB"}]
         * └─ TextNode: 'buy'
         *
         * WordNode[3] [data={"partOfSpeech":"CD"}]
         * ├─ TextNode: '5'
         * ├─ PunctuationNode: '.'
         * └─ TextNode: '2'
         *
         * WordNode[1] [data={"partOfSpeech":"NNS"}]
         * └─ TextNode: 'gallons'
         *
         * WordNode[1] [data={"partOfSpeech":"IN"}]
         * └─ TextNode: 'of'
         *
         * WordNode[1] [data={"partOfSpeech":"NN"}]
         * └─ TextNode: 'milk'
         */
    }
);
```

## API

None, the plugin automatically detects the part-of-speech tag for each [`WordNode`](https://github.com/wooorm/textom/tree/master#textomwordnode-nlcstwordnode) (using [dariusk/pos-js](https://github.com/dariusk/pos-js)), and stores the tag in `node.data.partOfSpeech`.

## Performance

On a MacBook Air, **retext** performs about 49% slower with **retext-pos**.


```text
           retext w/o retext-pos
  181 op/s » A paragraph (5 sentences, 100 words)
   20 op/s » A section (10 paragraphs, 50 sentences, 1,000 words)

           retext w/ retext-pos
   91 op/s » A paragraph (5 sentences, 100 words)
    9 op/s » A section (10 paragraphs, 50 sentences, 1,000 words)
```

## License

[MIT](LICENSE) © [Titus Wormer](http://wooorm.com)


History.md
<!--mdast setext-->

<!--lint disable no-multiple-toplevel-headings-->

<!--lint disable maximum-line-length-->

1.0.0 / 2015-08-01
==================

*   Refactor api for retext@1.0.0 (89ce36f)

0.2.1 / 2015-01-18
==================

*   Update changelog for 0.2.1 ([dc893af](https://github.com/wooorm/retext-pos/commit/dc893af))
*   Update `.gitignore` for bower ([9295897](https://github.com/wooorm/retext-pos/commit/9295897))
*   Update `.npmignore` for bower ([379fa02](https://github.com/wooorm/retext-pos/commit/379fa02))
*   Add Bower as an installation method in `Readme.md` ([03d3229](https://github.com/wooorm/retext-pos/commit/03d3229))
*   Add `bower.json` ([558ddb3](https://github.com/wooorm/retext-pos/commit/558ddb3))
*   Refactor `index.js` for updated installation methods ([a64e3a1](https://github.com/wooorm/retext-pos/commit/a64e3a1))
*   Update `.npmignore` for component ([6d27ac7](https://github.com/wooorm/retext-pos/commit/6d27ac7))
*   Update `.gitignore` for component ([94fbc76](https://github.com/wooorm/retext-pos/commit/94fbc76))
*   Add Component as an instalation method in `Readme.md` ([b426136](https://github.com/wooorm/retext-pos/commit/b426136))
*   Add `component.json` ([3fa175b](https://github.com/wooorm/retext-pos/commit/3fa175b))

0.2.0 / 2015-01-14
==================

*   Update changelog for 0.2.0 ([68576f4](https://github.com/wooorm/retext-pos/commit/68576f4))
*   Update copyright notice in `LICENSE` to include 2015 ([0cc8b77](https://github.com/wooorm/retext-pos/commit/0cc8b77))
*   Add link to whole license in `Readme.md` ([34512ba](https://github.com/wooorm/retext-pos/commit/34512ba))
*   Add Duo as an instalation method in `Readme.md` ([911d2c7](https://github.com/wooorm/retext-pos/commit/911d2c7))
*   Add support for duo ([d7a8d1e](https://github.com/wooorm/retext-pos/commit/d7a8d1e))
*   Add links to installation methods in `Readme.md` ([e91bf27](https://github.com/wooorm/retext-pos/commit/e91bf27))
*   Refactor fences code blocks in `Readme.md` ([60bc007](https://github.com/wooorm/retext-pos/commit/60bc007))
*   Refactor to adhere to strict jsdoc style ([3a18a69](https://github.com/wooorm/retext-pos/commit/3a18a69))
*   Add jscs-jsdoc configuration to `.jscs.json` ([958cfd6](https://github.com/wooorm/retext-pos/commit/958cfd6))
*   Add jscs-jsdoc as a dev-dependency ([2e80783](https://github.com/wooorm/retext-pos/commit/2e80783))
*   Refactor npm scripts for changes in npm ([43fcf8b](https://github.com/wooorm/retext-pos/commit/43fcf8b))
*   Update eslint ([889d215](https://github.com/wooorm/retext-pos/commit/889d215))
*   Refactor to use generic `changeinside` in new retext ([deb2937](https://github.com/wooorm/retext-pos/commit/deb2937))
*   Update retext to stable release ([1bf7542](https://github.com/wooorm/retext-pos/commit/1bf7542))

0.1.9 / 2014-12-06
==================

*   Update changelog for 0.1.9 ([3f72014](https://github.com/wooorm/retext-pos/commit/3f72014))
*   Update retext ([e7d11f0](https://github.com/wooorm/retext-pos/commit/e7d11f0))
*   Add benchmark ([b96a908](https://github.com/wooorm/retext-pos/commit/b96a908))
*   Add rewording to docs in `Readme.md` ([7473631](https://github.com/wooorm/retext-pos/commit/7473631))
*   Refactor npm script targets in `package.json` ([1a5a5ee](https://github.com/wooorm/retext-pos/commit/1a5a5ee))
*   Update eslint ([5a25b4e](https://github.com/wooorm/retext-pos/commit/5a25b4e))
*   Add link to personal website to copyright in `Readme.md` ([82b51e2](https://github.com/wooorm/retext-pos/commit/82b51e2))

0.1.8 / 2014-11-21
==================

*   Update changelog for 0.1.8 ([1a5cd7c](https://github.com/wooorm/retext-pos/commit/1a5cd7c))
*   Fix missing variable declaration in `test.js` ([410a686](https://github.com/wooorm/retext-pos/commit/410a686))
*   Refactor `Readme.md` ([6644541](https://github.com/wooorm/retext-pos/commit/6644541))
*   Add retext-inspect to dev-dependencies ([a205367](https://github.com/wooorm/retext-pos/commit/a205367))
*   Add flat badges to `Readme.md` ([b3cf7f6](https://github.com/wooorm/retext-pos/commit/b3cf7f6))
*   Fix incorrect executive rights on `test.js` ([d244879](https://github.com/wooorm/retext-pos/commit/d244879))
*   Refactor npm script targets in `package.json` ([9ce5b0e](https://github.com/wooorm/retext-pos/commit/9ce5b0e))
*   Add missing `.eslintrc` file ([f025dd0](https://github.com/wooorm/retext-pos/commit/f025dd0))
*   Update retext ([fdb2077](https://github.com/wooorm/retext-pos/commit/fdb2077))

0.1.7 / 2014-10-27
==================

*   Update changelog for 0.1.7 ([3c33ef5](https://github.com/wooorm/retext-pos/commit/3c33ef5))
*   Add `.eslintrc` ([244b8cb](https://github.com/wooorm/retext-pos/commit/244b8cb))
*   Refactor to disallow space after object keys ([ce6de65](https://github.com/wooorm/retext-pos/commit/ce6de65))
*   Refactor for changes in retext ([35a0077](https://github.com/wooorm/retext-pos/commit/35a0077))
*   Update eslint, mocha, retext, retext-visit, retext-content ([55cecfe](https://github.com/wooorm/retext-pos/commit/55cecfe))

0.1.6 / 2014-10-21
==================

*   Update changelog for 0.1.6 ([4798952](https://github.com/wooorm/retext-pos/commit/4798952))
*   Refactor test ([cf119ec](https://github.com/wooorm/retext-pos/commit/cf119ec))
*   Refactor .jscs.json ([51e4a87](https://github.com/wooorm/retext-pos/commit/51e4a87))
*   Update retext ([04fec58](https://github.com/wooorm/retext-pos/commit/04fec58))

0.1.5 / 2014-10-14
==================

*   Update changelog for 0.1.5 ([155a14d](https://github.com/wooorm/retext-pos/commit/155a14d))
*   Fix test for retext-visit@0.2.0 ([5445602](https://github.com/wooorm/retext-pos/commit/5445602))

0.1.4 / 2014-10-14
==================

*   Update changelog for 0.1.4 ([e7925e2](https://github.com/wooorm/retext-pos/commit/e7925e2))
*   Update retext, retext-visit ([ef32908](https://github.com/wooorm/retext-pos/commit/ef32908))

test.js
'use strict';

/* eslint-env mocha */

/*
 * Dependencies.
 */

var assert = require('assert');
var retext = require('retext');
var visit = require('unist-util-visit');
var pos = require('./');

/*
 * Fixtures.
 */

var sentence = 'I went to the store, to buy 5.2 gallons of milk.';

var tags = [
    'NN',
    'VBD',
    'TO',
    'DT',
    'NN',
    'TO',
    'VB',
    'CD',
    'NNS',
    'IN',
    'NN'
];

/*
 * Tests.
 */

describe('pos()', function () {
    retext().use(pos).process(sentence, function (err, file) {
        it('should work', function (done) {
            var index = -1;

            visit(file.namespace('retext').cst, 'WordNode', function (node) {
                assert.equal(node.data.partOfSpeech, tags[++index]);
            });

            done(err);
        });
    });
});


spec/retext-pos.spec.js
'use strict';

var pos,
    visit,
    content,
    Retext,
    assert,
    retext,
    tags,
    otherWords,
    otherTags;

/**
 * Module dependencies.
 */

pos = require('..');
visit = require('retext-visit');
content = require('retext-content');
Retext = require('retext');
assert = require('assert');

/**
 * Retext.
 */

retext = new Retext()
    .use(visit)
    .use(content)
    .use(pos);

/**
 * Fixtures.
 */

tags = ['DT', 'JJ', 'NNP', 'NN'];
otherWords = ['Another', 'harder', 'longer', 'paragraph'];
otherTags = ['DT', 'JJR', 'RB', 'NN'];

/**
 * Tests.
 */

describe('pos()', function () {
    it('should be a `function`', function () {
        assert(typeof pos === 'function');
    });

    it('should export an `attach` method', function () {
        assert(typeof pos.attach === 'function');
    });

    retext.parse('A simple, English, sentence', function (err, tree) {
        it('should not throw', function (done) {
            done(err);
        });

        it('should tag `WordNode`', function () {
            var index = -1;

            tree.visitType(tree.WORD_NODE, function (wordNode) {
                assert(wordNode.data.partOfSpeech === tags[++index]);
            });
        });

        it('should set `partOfSpeech` to `null` when `WordNode` ' +
            '(no longer?) has a value', function () {
                tree.visitType(tree.WORD_NODE, function (wordNode) {
                    wordNode.removeContent();

                    assert(wordNode.data.partOfSpeech === null);
                });
            }
        );

        it('should re-tag `WordNode`s when their value changes', function () {
            var index = -1;

            tree.visitType(tree.WORD_NODE, function (wordNode) {
                wordNode.replaceContent(otherWords[++index]);

                assert(wordNode.data.partOfSpeech === otherTags[index]);
            });
        });

        it('should set `partOfSpeech` to `null` when not attached or ' +
            'without value', function () {
                tree.visitType(tree.WORD_NODE, function (wordNode) {
                    wordNode.remove();

                    assert(wordNode.data.partOfSpeech !== null);

                    wordNode.removeContent();

                    assert(wordNode.data.partOfSpeech === null);
                });
            }
        );
    });
});


index.js
/**
 * @author Titus Wormer
 * @copyright 2014-2015 Titus Wormer
 * @license MIT
 * @module retext:pos
 * @fileoverview Retext part-of-speech (POS) tagger.
 */

'use strict';

/*
 * Dependencies.
 */

var visit = require('unist-util-visit');
var nlcstToString = require('nlcst-to-string');
var posjs;

/*
 * Duo and component / npm and component.
 */

try {
    posjs = require('pos');
} catch (err) {
    /* istanbul ignore next - browser */
    posjs = require('pos-js');
}

/*
 * Constants.
 */

var tagger = new posjs.Tagger();

/**
 * Patch a `partOfSpeech` property on `node`s.
 *
 * @param {NLCSTNode} node - Node.
 * @param {string} tag - Part-of-speech tag as returned
 *   by pos-js.
 */
function patch(node, tag) {
    var data = node.data || {};

    data.partOfSpeech = tag;

    node.data = data;
}

/**
 * Factory to gather parents and patch them based on their
 * childrens directionality.
 *
 * @return {function(node, index, parent)} - Can be passed
 *   to `visit`.
 */
function concatenateFactory() {
    var queue = [];

    /**
     * Gather a parent if not already gathered.
     *
     * @param {NLCSTWordNode} node - Word.
     * @param {number} index - Position of `node` in
     *   `parent`.
     * @param {NLCSTParentNode} parent - Parent of `child`.
     */
    function concatenate(node, index, parent) {
        if (parent && queue.indexOf(parent) === -1) {
            queue.push(parent);
        }
    }

    /**
     * Patch all words in `parent`.
     *
     * @param {NLCSTParentNode} node - Parent
     */
    function one(node) {
        var children = node.children;
        var length = children.length;
        var index = -1;
        var values = [];
        var words = [];
        var child;
        var tags;

        while (++index < length) {
            child = children[index];

            if (child.type === 'WordNode') {
                values.push(nlcstToString(child));
                words.push(child);
            }
        }

        tags = tagger.tag(values);
        index = -1;
        length = tags.length;

        while (++index < length) {
            patch(words[index], tags[index][1]);
        }
    }

    /**
     * Patch all parents.
     */
    function done() {
        var length = queue.length;
        var index = -1;

        while (++index < length) {
            one(queue[index]);
        }
    }

    concatenate.done = done;

    return concatenate;
}

/**
 * Patch `stem` on each node.
 *
 * @param {NLCSTNode} cst - Syntax tree.
 */
function transformer(cst) {
    var concatenate = concatenateFactory();

    visit(cst, 'WordNode', concatenate);

    concatenate.done();
}

/**
 * Define `metaphone`.
 *
 * @return {Function} - `transformer`.
 */
function attacher() {
    return transformer;
}

/*
 * Expose `metaphone`.
 */

module.exports = attacher;


benchmark.js
'use strict';

var Retext,
    pos;

/*
 * Dependencies.
 */

Retext = require('retext');
pos = require('./');

/*
 * Dependencies.
 */

var retext,
    retextWithPOS;

retext = new Retext();
retextWithPOS = new Retext().use(pos);

/*
 * Test data: A (big?) article (w/ 100 paragraphs, 500
 * sentences, 10,000 words);
 *
 * Source:
 *   http://www.gutenberg.org/files/10745/10745-h/10745-h.htm
 */

var sentence,
    paragraph,
    section;

sentence = 'Where she had stood was clear, and she was gone since Sir ' +
    'Kay does not choose to assume my quarrel.';

paragraph = 'Thou art a churlish knight to so affront a lady ' +
    'he could not sit upon his horse any longer. ' +
    'For methinks something hath befallen my lord and that he ' +
    'then, after a while, he cried out in great voice. ' +
    'For that light in the sky lieth in the south ' +
    'then Queen Helen fell down in a swoon, and lay. ' +
    'Touch me not, for I am not mortal, but Fay ' +
    'so the Lady of the Lake vanished away, everything behind. ' +
    sentence;

section = paragraph + Array(10).join('\n\n' + paragraph);

suite('retext w/o retext-pos', function () {
    bench('A paragraph (5 sentences, 100 words)', function (done) {
        retext.parse(paragraph, done);
    });

    bench('A section (10 paragraphs, 50 sentences, 1,000 words)',
        function (done) {
            retext.parse(section, done);
        }
    );
});

suite('retext w/ retext-pos', function () {
    bench('A paragraph (5 sentences, 100 words)', function (done) {
        retextWithPOS.parse(paragraph, done);
    });

    bench('A section (10 paragraphs, 50 sentences, 1,000 words)',
        function (done) {
            retextWithPOS.parse(section, done);
        }
    );
});
